extension/observer/endpointswatcher_test.go

```diff
@@ -20,76 +20,173 @@ import (
+	require.Equal(t, expected, existingEndpoints(t, watcher, notify.ID()))
+
+	lister.updateEndpoint(2, "updated_target")
+	notify.numToBeChanged = 1
 
-	ml.updateEndpoint(2, "updated_target")
-	ew.refreshEndpoints(mn)
+	watcher.notifyOfEndpoints(notify.ID())
 
 	expected["2"] = Endpoint{ID: "2", Target: "updated_target"}
-	require.Equal(t, expected, ew.existingEndpoints)
+	require.Equal(t, expected, existingEndpoints(t, watcher, notify.ID()))
+
+	watcher.Unsubscribe(notify)
+	require.Nil(t, existingEndpoints(t, watcher, notify.ID()))
 }
 
-func setup() (*mockEndpointsLister, EndpointsWatcher, mockNotifier) {
-	ml := &mockEndpointsLister{
-		endpointsMap: map[EndpointID]Endpoint{},
+func TestNotifyOfEndpointsMultipleNotify(t *testing.T) {
+	lister, watcher, notifyOne := setup(t)
+	defer watcher.StopListAndWatch()
+
+	lister.addEndpoint(0)
+	notifyOne.numToBeAdded = 1
+	watcher.ListAndWatch(notifyOne)
+
+	notifyTwo := newMockNotififier(t, "notify2")
+	lister.addEndpoint(1)
+	notifyTwo.numToBeAdded = 2
+	watcher.ListAndWatch(notifyTwo)
+
+	expectedOne := map[EndpointID]Endpoint{"0": {ID: "0"}}
+	require.Equal(t, expectedOne, existingEndpoints(t, watcher, notifyOne.ID()))
+
+	expectedTwo := map[EndpointID]Endpoint{"0": {ID: "0"}, "1": {ID: "1"}}
+	require.Equal(t, expectedTwo, existingEndpoints(t, watcher, notifyTwo.ID()))
+
+	lister.addEndpoint(2)
+	lister.addEndpoint(3)
+	notifyOne.numToBeAdded = 3
+	notifyTwo.numToBeAdded = 2
+	lister.removeEndpoint(0)
+	delete(expectedOne, "0")
+	delete(expectedTwo, "0")
+	notifyOne.numToBeRemoved = 1
+	notifyTwo.numToBeRemoved = 1
+
+	watcher.notifyOfEndpoints(notifyOne.ID(), notifyTwo.ID())
+
+	expectedOne["1"] = Endpoint{ID: "1"}
+	expectedOne["2"] = Endpoint{ID: "2"}
+	expectedOne["3"] = Endpoint{ID: "3"}
+	expectedTwo["2"] = Endpoint{ID: "2"}
+	expectedTwo["3"] = Endpoint{ID: "3"}
+
+	require.Equal(t, expectedOne, existingEndpoints(t, watcher, notifyOne.ID()))
+	require.Equal(t, expectedTwo, existingEndpoints(t, watcher, notifyTwo.ID()))
+
+	lister.updateEndpoint(2, "updated_target")
+	notifyOne.numToBeChanged = 1
+	notifyTwo.numToBeChanged = 1
+	watcher.notifyOfEndpoints(notifyOne.ID(), notifyTwo.ID())
+
+	expectedOne["2"] = Endpoint{ID: "2", Target: "updated_target"}
+	require.Equal(t, expectedOne, existingEndpoints(t, watcher, notifyOne.ID()))
+	expectedTwo["2"] = Endpoint{ID: "2", Target: "updated_target"}
+	require.Equal(t, expectedTwo, existingEndpoints(t, watcher, notifyTwo.ID()))
+
+	watcher.Unsubscribe(notifyOne)
+	require.Nil(t, existingEndpoints(t, watcher, notifyOne.ID()))
+
+	lister.addEndpoint(4)
+	notifyTwo.numToBeAdded = 1
+	watcher.ListAndWatch(notifyTwo)
+	expectedTwo["4"] = Endpoint{ID: "4"}
+	require.Equal(t, expectedTwo, existingEndpoints(t, watcher, notifyTwo.ID()))
+
+	watcher.Unsubscribe(notifyTwo)
+	require.Nil(t, existingEndpoints(t, watcher, notifyTwo.ID()))
+}
+
+func existingEndpoints(t *testing.T, watcher *EndpointsWatcher, id NotifyID) map[EndpointID]Endpoint {
```

Comment:
For test helpers, could I ask that you use `tb testing.TB` ? 

The reason being that it allows code reuse within benchmarking and fuzzing.

Reply:
updated